### PR TITLE
Shareable cipher design

### PR DESCRIPTION
**Pull request by rachelli99**
restructured the layout for crack cipher

### CODE REVIEW

**Comment by mattxwang** [comment]
Hey Rachel, glad you're back with us! Hm, I think you're still tripping up a few ESlint errors - can you fix those first @rachelli99?

(also, do you mind merging master back into your branch first?)

File: src/Components/ShareCipher/Congrats.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import React from 'react';
+import Anime from 'react-anime';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faGraduationCap } from '@fortawesome/free-solid-svg-icons';
+
+
+function Congrats(props) {
+
+    let plaintext = props.cipher[1];               // plaintext of the ciphered message
+    let ciphertext = props.cipher[0];             // current guess at the deciphered message
```
Comment (krashanoff) [inline src/Components/ShareCipher/Congrats.js]: Since our props don't change within the scope of the component, these can be `const`!

File: src/Components/ShareCipher/Congrats.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    // load the described cipher.
+    
+    return (
+        <div className="container center">
+            <h1 className='title is-size-2 underline mt-4'>Are you the new crack master?</h1>
```
Comment (krashanoff) [inline src/Components/ShareCipher/Congrats.js]: A small change of wording would be nice here. "Crack master" is probably not the term we want to throw around with kids 😅

File: src/Components/ShareCipher/Congrats.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                <section className='share-section'>
+                    <Anime 
+                    easing="linear"
+                    delay='100' 
+                    opacity={['0%','100%']}
+                    duration="1000"
+                    >
+                    <FontAwesomeIcon icon={faGraduationCap} alt='key' size="6x" color="#FFB800" />
+                        <p className='is-size-3 mb-5'>Congratulations! </p>
+                    </Anime>
```
Comment (krashanoff) [inline src/Components/ShareCipher/Congrats.js]: This is rather picky but I'd rather you indent this properly. It helps significantly with readability. If you want to automate this, [there are linters that can do this](https://prettier.io/docs/en/precommit.html)

```suggestion
                <section className='share-section'>
                    <Anime 
                        easing="linear"
                        delay='100' 
                        opacity={['0%','100%']}
                        duration="1000"
                    >
                        <FontAwesomeIcon icon={faGraduationCap} alt='key' size="6x" color="#FFB800" />
                        <p className='is-size-3 mb-5'>Congratulations! </p>
                    </Anime>
```

File: src/Components/ShareCipher/CrackCipher.js [hunk 4]
```diff
@@ -1,128 +1,185 @@
...
-    // if provided a hash, then try to load it.
-    if (hash && !getStatus && !ciphertext)
-        loadCurrentHash();
+    const shiftResult = caesarShift(ciphertext ? ciphertext : '', shamt) === (plaintext? plaintext : '').toUpperCase();
+    const shiftText = caesarShift((ciphertext ? ciphertext : '').toUpperCase(), shamt);
 
-    // If the user navigated directly to this page
-    if (plaintext === null) {
+    // if the user was provided a valid hash
+    if(firstTimeLoad){  // fade in effect on first time load, instead of everytime the wheel turns
```
Comment (krashanoff) [inline src/Components/ShareCipher/CrackCipher.js]: This is a *really* big if/else statement. Instead of using this, how about trying something a little more compact?

Here's an example: here, I do the same thing: use a state variable to track whether the page has been loaded in already. However, instead of using an if/else with code duplication, I instead use an if/else before the render to control whether the `<Anime />` tags do anything useful.

```js
import React, { useState } from 'react';
import Anime from 'react-anime';

export default function App() {
  const [ hasPlayed, setHasPlayed ] = useState(false);  // Tracking animation plays.
  const [ dummyEvent, setDummyEvent ] = useState(true);    // Dummy event to force rerenders.

  return (
    <div className="App">

      {/* Use a ternary to control whether the animation does anything useful. */}
      <Anime opacity={!hasPlayed ? [0,1] : [1]} duration={5000}>
        <h1>Hi</h1>
      </Anime>

      <button onClick={() => setDummyEvent(!dummyEvent)}>This will play it again</button>
      <button onClick={() => setHasPlayed(true)}>This will stop it from animating</button>
    </div>
  );
}
```

File: src/Components/ShareCipher/shareCipher.scss [hunk 5]
```diff
@@ -14,3 +14,10 @@
 .create-link * {
     transition: all .3s ease-out;
 }
+
+.center{
+    justify-content: center;
+    align-content: center;
+    align-items: center;
+    text-align: center;
+}
```
Comment (krashanoff) [inline src/Components/ShareCipher/shareCipher.scss:23]: Remember to set this container to `display: flex;`!

File: src/Components/ShareCipher/SolveCipher.js [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                        <h1 className='title is-size-2 underline mt-4'>Can you solve my cipher?</h1>
+            
+                            <section className='share-section'>
+                                <p className='is-size-4 mb-3'>Your friend has encoded a cipher for you to crack, </p>
+                                <p className='is-size-4 mb-3'>the cipher is encrypted with <u>caesar</u> encoding. </p>
+                                <p className='is-size-4 mb-3'>Every letter in the ciphertext is shifted by a constant amount. </p>
+            
+                                <div className="columns is-vcentered is-size-3">
+                                <div class="column"></div>
+                                <div class="column"></div>
```
Comment (krashanoff) [inline src/Components/ShareCipher/SolveCipher.js]: How come we have these extra columns here? If we are using them for formatting, it may pay dividends to simply leave them out and let flexbox do its thing.

File: src/Components/ShareCipher/SolveCipher.js [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+                                <div class="column" style={{color:'#CC0000'}}>
+                                <Anime easing="linear"
+                                loop={true}
+                                duration={8000}
+                                opacity={['0%','100%','0%','100%','0%','0%','0%','0%','0%','0%','0%','0%','0%','0%']}
+                                > P </Anime>
+                                </div>
```
Comment (krashanoff) [inline src/Components/ShareCipher/SolveCipher.js]: Two things:
1. Formatting here will help for readability
2. I noticed a lot of duplication here. Instead of creating the same lines of code over and over again, let's leverage JSX to generate them flexibly. Here's an example:

```js
import React from 'react';
import Anime from 'react-anime';

// Generates some animated components from an input string.
// Each word is transformed into an h2 tag wrapped with an
// <Anime /> component.
export default function App() {
  // create some array to work with.
  const someArr = 'A long string of words split into headers'.split(' ');

  return (
    <div>
      {/* for each word, map it to our desired template. */}
      {
        someArr.map(word => (
          <Anime opacity={[0,1]} duration={5000}>
            <h2 className='word'>{word}</h2>
          </Anime>
        ))
      }
    </div>
  );
}
```

This renders:

image

File: src/Components/ShareCipher/SolveCipher.js [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+                        
+                </Anime>
+
+                <div className='my-5'></div>
+            </Route>
+
+            <Route path={match.path}>
+                <section className='container'>
+
+                    <Anime easing="linear" duration="700" loop={false} opacity={['0%','100%']} >
```
Comment (krashanoff) [inline src/Components/ShareCipher/SolveCipher.js]: You will likely need to repeat what you did in your crackCipher component to track whether this animation has played out, otherwise you will get flashing when you shift the wheel.

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Rachel! Thanks a ton for tackling this issue - I know it is rather chunky.

I really appreciated all the animations that were added in to the process, and *loved* the animation you used to explain the cipher shifting process. Along with the spacing and dividers, things are looking much cleaner. This said, there are still a good many details that need to be tackled. They are mostly small things like ensuring that the width of the container doesn't change, or that the page doesn't flash on cipher load.

I have left code-specific comments in the thread attached to this review, but here are the more general ones. I can provide specific points in the code to change to address them upon request:

## General

It would be nice to use a container with a minimum or set width to prevent the weird changes in container width that we see when inputting text. Take a look at the below screenshots for an example:

image

image

Notice how the width changes? We can prevent this by keeping the inner container to a wide minimum width:

```css
.cipher-container {
    min-width: 90%;
}
```

## Can you solve my cipher?

image

I'd like for the animation here to play out a single time, then stop to reveal the entire process to the user. In its current state, it gets the point across, but it can be a little hard to follow. On a more specific note, I think that the divider should have a little bit more spacing to keep it from colliding with the letters.

In this same section, when we input the link for the cipher, the entire page flashes.

After we paste it, I think that we should be able to just press "Load" and then immediately have the cipher solving portion come up, rather than manually clicking the "Next" that appears.

## Let's get cracking!

image

It would be nice to see the decoded text somewhere as well so that way the message gets through to the user :+1:

Sorry for the huge review! Let me know if you have any questions.

File: src/Components/ShareCipher/shareCipher.scss [hunk 9]
```diff
@@ -14,3 +14,15 @@
...
+
+.cipher-container {
+    min-width: 60vw;
+    max-width: 60vw;
```
Comment (krashanoff) [inline src/Components/ShareCipher/shareCipher.scss:27]:
```suggestion
    width: 60vw;
```

You can use `width: 60vw` here :+1:

**Review by krashanoff — APPROVED** [review]
Hey Rachel, sorry for the delay on this PR. Thank you for addressing my feedback, great work! I think I'd like to merge this real quick just to get the visuals up to snuff. Some comments on things that we should change going forward:
* The appearance on mobile is a little strange when solving a cipher.
* There's a slight visual bug I encountered on the `SolveCipher` animation. On load, the component flashes all components for a split second.
* We could clean up the styles.

Aside from this, things are looking much more conducive to the vision I had setting out with the cipher project. Nice work Rachel!

**Comment by krashanoff** [comment]
Going to ping @rachelli99 one more time for confirmation. Otherwise, will merge for sake of time.

**Comment by rachelli99** [comment]
> Going to ping @rachelli99 one more time for confirmation. Otherwise, will merge for sake of time.

Sounds good, feel free to merge it. I'll work on the other minor fixes as we go. Thanks for all the feedbacks and help!
